File: pc02/pc02_02/c/pc02_02.c
/**
 * Poker hands
 * pc110202
 * uva10315
 */

#include <stdio.h>
#include <string.h>

#define BUFSIZE 40

typedef struct {
	int value;
	int suit;
} card_t;

int char_to_value(char c) {
	static char *values = "23456789TJQKA";
	for (int i = 0; values[i] != '\0'; i++) {
		if (values[i] == c) {
			return i;
		}
	}
	return -1;
}

int char_to_suit(char c) {
	static char *suits = "CDHS";
	for (int i = 0; suits[i] != '\0'; i++) {
		if (suits[i] == c) {
			return i;
		}
	}
	return -1;
}

/* Parse 10 cards from input line into black[5] and white[5] */
void parse_inputs(const char *input, card_t *black, card_t *white) {
	/* Each i is a value for a card, black or white */
	for (int i = 0; i < 5; i++) {
		int value_char;
		int suit_char;

		/* First the black cards */
		value_char = input[i * 3];
		suit_char = input[i * 3 + 1];

		black[i].value = char_to_value(value_char);
		black[i].suit = char_to_suit(suit_char);

		/* Then white cards */
		value_char = input[15 + i * 3];
		suit_char = input[15 + i * 3 + 1];

		white[i].value = char_to_value(value_char);
		white[i].suit = char_to_suit(suit_char);
	}
}

/* Bubble sort with early stopping */
void sort_hand(card_t *hand) {
	int changed;

	while (1) {
		changed = 0;

		for (int i = 0; i < 4; i++) {
			if (hand[i].value > hand[i + 1].value) {
				card_t tmp = hand[i];
				hand[i] = hand[i + 1];
				hand[i + 1] = tmp;
				changed = 1;
			}
		}

		if (!changed) {
			break;
		}
	}
}

int is_flush(card_t *hand) {
	int suit = hand[0].suit;
	for (int i = 1; i < 5; i++) {
		if (hand[i].suit != suit) {
			return 0;
		}
	}
	return 1;
}

int is_straight(card_t *hand) {
	for (int i = 1; i < 5; i++) {
		if (hand[i].value != hand[i - 1].value + 1) {
			if (i == 4 &&
				hand[0].value == 0 &&
				hand[1].value == 1 &&
				hand[2].value == 2 &&
				hand[3].value == 3 &&
				hand[4].value == 12) {
				return 1;
			}
			return 0;
		}
	}
	return 1;
}


void get_value_counts(const card_t *hand, int *counts) {
	for (int i = 0; i < 13; i++) {
		counts[i] = 0;
	}

	for (int i = 0; i < 5; i++) {
		counts[hand[i].value]++;
	}
}

/* Classify sorted rank, fill tiebreakers */
int classify_hand(card_t *hand, int *tiebreakers) {
	int counts[13];
	get_value_counts(hand, counts);

	int pairs = 0;
	int three = 0;
	int four = 0;

	int pair_values[2];
	int three_value = -1;
	int four_value = -1;

	for (int value = 0; value < 13; value++) {
		if (counts[value] == 4) {
			four_value = value;
		} else if (counts[value] == 3) {
			three_value = value;
		} else if (counts[value] == 2) {
			pair_values[pairs++] = value;
		}
	}

	int flush = is_flush(hand);
	int straight = is_straight(hand);
	int highest = hand[4].value;	/* This assumes sorted hand */

	if (straight && flush) {
		tiebreakers[0] = highest;
		return 8;
	}

	if (four_value != -1) {
		tiebreakers[0] = four_value;
		return 7;
	}

	if (three_value != -1 && pairs == 1) {
		tiebreakers[0] = three_value;
		return 6;
	}

	if (flush) {
		for (int i = 0; i < 5; i++) {
			tiebreakers[i] = hand[4 - i].value;
		}
		return 5;
	}

	if (straight) {
		tiebreakers[0] = highest;
		return 4;
	}

	if (three_value != -1) {
		tiebreakers[0] = three_value;
		return 3;
	}

	if (pairs == 2) {
		if (pair_values[0] > pair_values[1]) {
			tiebreakers[0] = pair_values[0];
			tiebreakers[1] = pair_values[1];
		} else {
			tiebreakers[0] = pair_values[1];
			tiebreakers[1] = pair_values[0];
		}
		for (int i = 4; i >= 0; i--) {
			if (hand[i].value != pair_values[0] && hand[i].value != pair_values[1]) {
				tiebreakers[2] = hand[i].value;
				break;
			}
		}
		return 2;
	}

	if (pairs == 1) {
		tiebreakers[0] = pair_values[0];
		int idx = 1;
		for (int i = 4; i >= 0; i--) {
			if (hand[i].value != pair_values[0]) {
				tiebreakers[idx++] = hand[i].value;
			}
		}
		return 1;
	}

	for (int i = 0; i < 5; i++) {
		tiebreakers[i] = hand[4 - i].value;
	}
	return 0;
}

int compare_hands(card_t *black, card_t *white) {
	int b_tie[5] = {0};
	int w_tie[5] = {0};

	int b_rank = classify_hand(black, b_tie);
	int w_rank = classify_hand(white, w_tie);

	if (b_rank < w_rank) {
		return 1;
	}

	if (b_rank > w_rank) {
		return -1;
	}

	for (int i = 0; i < 5; i++) {
		if (b_tie[i] > w_tie[i]) {
			return 1;
		}

		if (b_tie[i] < w_tie[i]) {
			return -1;
		}
	}

	return 0;
}

int main(void) {
	char input[BUFSIZE];

	while (fgets(input, BUFSIZE, stdin) != NULL) {
		char *last = strchr(input, '\n');
		if (last != NULL) {
			*last = '\0';
		}

		card_t black[5];
		card_t white[5];
		parse_inputs(input, black, white);

		sort_hand(black);
		sort_hand(white);

		int result = compare_hands(black, white);

		switch (result) {
			case 1:
				printf("Black wins.\n");
				break;
			case -1:
				printf("White wins.\n");
				break;
			default:
				printf("Tie.\n");
		}
	}

	return 0;
}

File: pc02/pc02_02/python/pc02_02.py
"""
Poker hands
pc110202
uva10315
"""


class Card:
    def __init__(self, value, suit):
        self.value = value
        self.suit = suit


def char_to_value(c):
    """Convert card value character to integer."""
    values = "23456789TJQKA"
    return values.index(c)


def char_to_suit(c):
    """Convert card suit character to integer."""
    suits = "CDHS"
    return suits.index(c)


def parse_inputs(input_str):
    """Parse 10 cards from input string into black[5] and white[5]."""
    black = []
    white = []

    for i in range(5):
        black.append(Card(char_to_value(input_str[i * 3]),
                         char_to_suit(input_str[i * 3 + 1])))
        white.append(Card(char_to_value(input_str[15 + i * 3]),
                         char_to_suit(input_str[15 + i * 3 + 1])))

    return black, white


def sort_hand(hand):
    """Sort hand by card value."""
    hand.sort(key=lambda card: card.value)


def is_flush(hand):
    """Check if hand is a flush."""
    suit = hand[0].suit
    for i in range(1, 5):
        if hand[i].suit != suit:
            return False
    return True


def is_straight(hand):
    """Check if hand is a straight."""
    for i in range(1, 5):
        if hand[i].value != hand[i - 1].value + 1:
            if (i == 4 and
                hand[0].value == 0 and
                hand[1].value == 1 and
                hand[2].value == 2 and
                hand[3].value == 3 and
                hand[4].value == 12):
                return True
            return False
    return True


def get_value_counts(hand):
    """Get count of each card value."""
    counts = [0] * 13
    for card in hand:
        counts[card.value] += 1
    return counts


def classify_hand(hand):
    """Classify hand and return (rank, tiebreakers)."""
    counts = get_value_counts(hand)

    pairs = []
    three_value = -1
    four_value = -1

    for value in range(13):
        if counts[value] == 4:
            four_value = value
        elif counts[value] == 3:
            three_value = value
        elif counts[value] == 2:
            pairs.append(value)

    flush = is_flush(hand)
    straight = is_straight(hand)
    highest = hand[4].value

    tiebreakers = [0] * 5

    if straight and flush:
        tiebreakers[0] = highest
        return 8, tiebreakers

    if four_value != -1:
        tiebreakers[0] = four_value
        return 7, tiebreakers

    if three_value != -1 and len(pairs) == 1:
        tiebreakers[0] = three_value
        return 6, tiebreakers

    if flush:
        for i in range(5):
            tiebreakers[i] = hand[4 - i].value
        return 5, tiebreakers

    if straight:
        tiebreakers[0] = highest
        return 4, tiebreakers

    if three_value != -1:
        tiebreakers[0] = three_value
        return 3, tiebreakers

    if len(pairs) == 2:
        if pairs[0] > pairs[1]:
            tiebreakers[0] = pairs[0]
            tiebreakers[1] = pairs[1]
        else:
            tiebreakers[0] = pairs[1]
            tiebreakers[1] = pairs[0]

        for i in range(4, -1, -1):
            if hand[i].value != pairs[0] and hand[i].value != pairs[1]:
                tiebreakers[2] = hand[i].value
                break

        return 2, tiebreakers

    if len(pairs) == 1:
        tiebreakers[0] = pairs[0]
        idx = 1
        for i in range(4, -1, -1):
            if hand[i].value != pairs[0]:
                tiebreakers[idx] = hand[i].value
                idx += 1

        return 1, tiebreakers

    for i in range(5):
        tiebreakers[i] = hand[4 - i].value

    return 0, tiebreakers


def compare_hands(black, white):
    """Compare two hands and return winner."""
    b_rank, b_tie = classify_hand(black)
    w_rank, w_tie = classify_hand(white)

    if b_rank > w_rank:
        return "Black"

    if b_rank < w_rank:
        return "White"

    for i in range(5):
        if b_tie[i] > w_tie[i]:
            return "Black"

        if b_tie[i] < w_tie[i]:
            return "White"

    return "Tie"


def main():
    import sys
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        black, white = parse_inputs(line)
        sort_hand(black)
        sort_hand(white)

        result = compare_hands(black, white)

        if result == "Black":
            print("Black wins.")
        elif result == "White":
            print("White wins.")
        else:
            print("Tie.")


if __name__ == '__main__':
    main()
